packages/jest-haste-map/src/index.js

```diff
@@ -681,6 +706,49 @@ class HasteMap extends EventEmitter {
     });
   }
 
+  /**
+   * This function should be called when the file under `filePath` is removed
+   * or changed. When that happens, we want to figure out if that file was
+   * part of a group of files that had the same ID. If it was, we want to
+   * remove it from the group. Furthermore, if there is only one file
+   * remaining in the group, then we want to restore that single file as the
+   * correct resolution for its ID, and cleanup the duplicates index.
+   */
+  _recoverDuplicates(
+    hasteMap: InternalHasteMap,
+    filePath: string,
+    moduleName: string,
+  ) {
+    let dupsByPl = hasteMap.duplicates[moduleName];
+    if (dupsByPl == null) {
+      return;
+    }
+    const platform =
+      getPlatformExtension(filePath, this._options.platforms) ||
+      H.GENERIC_PLATFORM;
+    let dups = dupsByPl[platform];
+    if (dups == null) {
+      return;
+    }
+    dupsByPl = hasteMap.duplicates[moduleName] = (copy(dupsByPl): any);
+    dups = dupsByPl[platform] = (copy(dups): any);
+    const dedupType = dups[filePath];
+    delete dups[filePath];
```

Comment:
Look at the code above the callsite of `_recoverDuplicates`. What we do is that we remove all references to the changed code, both in the `files` index and in the `moduleMap`, so that when we process it we know we just have to insert it, there's nothing else to remove elsewhere.

Here is the same idea. Basically as soon as a file changed, whether deleted/changed, we remove it from the duplication index as if it had been deleted. Later on, we run the file-processing code above, that will re-create the duplication index for that file in case the ID conflict still happens. Of course that means if a conflicted file is changed again and again, we'll delete and re-create the duplication index for its ID again and again. But, this makes the logic simpler and more consistent with the rest of the `jest-haste-map` architecture. We could implement a complicated logic that checks if the ID changed, and delete the file from the duplication index only if it did change. But, that logic would be more complex, and more importantly, more fragile. It'd also be harder to implement given the current structure of the system ^_^

Does that answer the question?